### ISSUE TITLE
Terribly slow conditional updates

### ISSUE THREAD

**Issue by phquanta** [tensorflow/tensorflow#26841]
I'm trying to implement conditional  tensor updates such as:

#######################################################
Given i:

tf.scatter_update(self.vOut,i,[tf.cond( tf.less(tf.cast(tf.random_uniform([1])[0],tf.float64),tf.minimum(tf.cast(1.,tf.float64),  tf.exp(-self.dUk(vv,hh,vIn,hIn,i,j,krepl)))),
                 lambda:tf.slice(vv,[j,0],[1,self.n_v])[0],lambda:tf.slice(vIn,[i,j,0],[1,1,self.n_v])[0][0]) for j in kk])

 def U(self,x,y,x1,y1):
            return tf.multiply(tf.cast(self.P1,tf.float64),tf.exp(-((tf.norm(x-y)+tf.norm(x1-y1))/self.P2)**2.))

  def dU(self,x,y,v,z,x1,y1,v1,z1):    
          return    tf.subtract(self.U(x,y,x1,y1),self.U(v,z,v1,z1))

 def dUk(self,vv,hh,vIn,hIn,i,j,krepl):    
          return    tf.reduce_sum([tf.cond(tf.not_equal(p,i),
             lambda: self.dU(tf.slice(vv,[j,0],[1,self.n_v])[0],tf.slice(vIn,[p,j,0],[1,1,self.n_v])[0][0], 
             tf.slice(vIn,[i,j,0],[1,1,self.n_v])[0][0],tf.slice(vIn,[p,j,0],[1,1,self.n_v])[0][0],
             tf.slice(hh,[j,0],[1,self.n_h])[0],tf.slice(hIn,[p,j,0],[1,1,self.n_h])[0][0],
             tf.slice(hIn,[i,j,0],[1,1,self.n_h])[0][0],tf.slice(hIn,[p,j,0],[1,1,self.n_h])[0][0]
             ),
             lambda: tf.cast(0., tf.float64)) for p in  krepl])
 #############################################################

The problem i'm encountering is being  terribly slow compared to just numpy computation, i.e.

pure dUk computation:
dUk time (TF): 2.5306854248046875
dUk time (Numpy) time: 0.0

Update computation:
Time for Scatter (TF): 12.936962366104126
Time for Scatter (Numpy): 0.015621185302734375

Everything is done on CPU.

Is it normal to be that slow ?

**Comment by jvishnuvardhan** [comment]
@phquanta Could you provide system details and TF version [here](https://github.com/tensorflow/tensorflow/issues/new?template=00-bug-performance-issue.md). Thanks!

**Comment by phquanta** [comment]
Tensorflow: version 1.7.0 (CPU based installation)
Python: 3.6
OS: Windows 8.1
OS Name	Microsoft Windows 8.1
Version	6.3.9600 Build 9600
Other OS Description 	Not Available
System Model	XPS 8700
Processor	Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz, 3601 Mhz, 4 Core(s), 8 Logical 
Installed Physical Memory (RAM)	24.0 GB
GPU: Nvidia GTX 1050Ti
IDE: Spyder 3.3.1


Aslo this operation

j=1

aa=self.sess.run(tf.map_fn(lambda x: (tf.tensordot(x[0],x[1],0)),elems=tf.convert_to_tensor(self.vt[:,j,:]), tf.convert_to_tensor(self.ht[:,j,:])),dtype=tf.float64))

is twice as slow compared  to:

[np.outer(self.vt[i,j,:],self.ht[i,j,:])*expF[i] for i in self.lRepls]


self.lRepls=[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]
np.shape(self.vt)=(16,10,784)
np.shape(self.ht)=(16,10,340)

**Comment by jvishnuvardhan** [comment]
@phquanta Could you check with latest version of TF and check whether the bug persists? Thanks!

**Comment by phquanta** [comment]
Thanks. Will do.

**Comment by phquanta** [comment]
Did check, it still persists. 

Numpy:
start=time.time()
self.wNegJ[j,:,:]+=np.add.reduce([np.outer(self.vt[i,j,:],self.ht[i,j,:])*expF[i] for i in self.lRepls],axis=0)                        
print("Time for Average step0:",time.time()-start) 

Time for Average step0: 0.26468753814697266

Tensorflow:
               bb1=tf.convert_to_tensor(self.vt[:,j,:])
               bb2=tf.convert_to_tensor(self.ht[:,j,:])
                        star1t=time.time()
                        with tf.device(self.device):  
                         
                         aa=self.sess.run(tf.map_fn(lambda x: (tf.tensordot(x[0],x[1],0)),elems=(bb1,
                                                                                bb2),dtype=tf.float64))
                        star1t=time.time()

                        self.wNegJ[j,:,:]+=np.add.reduce([aa[i]*expF[i]
                                    for i in self.lRepls],axis=0)                        
                        
                                      print("Time for Average step0:",time.time()-star1t)   


Time for Average step0: 0.4763360023498535

**Comment by phquanta** [comment]
Sorry, TF version is 1.11.0

**Comment by mohantym** [comment]
Hi @phquanta ,
We see that you are using old version of Tensorflow which is officially considered as end of life, We recommend that you upgrade the code base to 2.6 version and let us know if the issue still persists in newer versions .Please open a new issue in case you face any errors, we will get you the right help .Thanks!
